Correct detection of WSDL files

theamanrawat opened the pull request:
As @kingthorin told.

```
Should be as simple as:

- return baseURL.endsWith(".wsdl")
-                    || contentType.equals("text/xml")
-                    || contentType.equals("application/wsdl+xml");
+ return (baseURL.endsWith(".wsdl")
+                    && contentType.equals("text/xml"))
+                    || contentType.equals("application/wsdl+xml");
Or

- return baseURL.endsWith(".wsdl")
-                    || contentType.equals("text/xml")
-                    || contentType.equals("application/wsdl+xml");
+ return baseURL.endsWith(".wsdl")
+                    && (contentType.equals("text/xml")
+                    || contentType.equals("application/wsdl+xml"));
```

 I made changes.

Fixes zaproxy/zaproxy/issues/6440

Signed-off-by: Aman Rawat <rawataman6525@gmail.com>

psiinon commented:
Can you update the unit test please :)
https://github.com/zaproxy/zap-extensions/blob/master/addOns/soap/src/test/java/org/zaproxy/zap/extension/soap/WSDLFilePassiveScanRuleTestCase.java

theamanrawat commented:
Yeah sure I'll do that

thc202 commented:
Please update the changelog.

thc202 commented on addOns/soap/src/main/java/org/zaproxy/zap/extension/soap/WSDLFilePassiveScanRule.java:
```diff
@@ -57,8 +57,7 @@ public boolean isWsdl(HttpMessage msg) {
             String contentType = header.getHeader(HttpHeader.CONTENT_TYPE).trim();
-            return baseURL.endsWith(".wsdl")
-                    || contentType.equals("text/xml")
+            return (baseURL.endsWith(".wsdl") && contentType.equals("text/xml"))
```
Why just ending with `.wsdl` is no longer considered WSDL?

kingthorin replied:
That’s why I included two different groupings.

It can be any URL ending in wsdl with xml content type or wsdl/xml content type. Or, only URLs ending in wsdl with either content type.

kingthorin replied:
Or something I didn’t consider I suppose

ricekot replied:
I think that the `"text/xml"` check can be omitted altogether and the check should just be
```java
baseURL.endsWith(".wsdl") || contentType.equals("application/wsdl+xml")
```

thc202 replied:
Sounds good to me.

theamanrawat replied:
> I think that the `"text/xml"` check can be omitted altogether and the check should just be
> 
> ```java
> baseURL.endsWith(".wsdl") || contentType.equals("application/wsdl+xml")
> ```

I added this and Now I think that `ShouldNotAlertWithJustXmlContentType` in unit test does not make any sense so should I remove it? 

In unit test I added three cases:

1. Should alert when file found with wsdl extension.
2. Should alert when wsdl content type found.
3. Should not alert when file extension is not wsdl.

thc202 replied:
Yes, that test can be removed now.

theamanrawat commented:
I added a unit test and updated changelog

Previously, It was giving an alert in a case where file is not ending with .wsdl and it's content type is text/xml and Now it's not

thc202 commented on addOns/soap/src/test/java/org/zaproxy/zap/extension/soap/Sample.java:
```diff
@@ -172,4 +172,14 @@ public static HttpMessage setSoapVersionTwoRequest(HttpMessage msg)
         return msg;
     }
+
+    public static HttpMessage setXmlResponseHeader(HttpMessage msg)
```
Better add this in the required test method.

thc202 commented on addOns/soap/CHANGELOG.md:
```diff
@@ -26,6 +26,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 - Fix exception with Java 9+ (Issue 4037).
 - SOAP operations are no longer overwritten in sites tree (Issue 1867).
 - Persist the add-on configuration required by the scan rules in the ZAP database (Issue 4866).
+- Fix detection of WSDL files and added unit test (Issue 6440).
```
`Fix detection of WSDL files (Issue 6440).`, the unit test part is not relevant for the end user.

thc202 commented on addOns/soap/src/test/java/org/zaproxy/zap/extension/soap/WSDLFilePassiveScanRuleTestCase.java:
```diff
@@ -58,4 +59,15 @@ public void isWsdlTest()
     }
+
+    @Test
+    public void isXmlContentType() throws IOException {
```
`shouldNotBeWsdlWithJustTextXmlContentType`, or something like that.

Could you add tests covering the other cases?

theamanrawat commented:
@thc202, I've done all the requested changes.

ricekot commented on addOns/soap/src/test/java/org/zaproxy/zap/extension/soap/WSDLFilePassiveScanRuleTestCase.java:
```diff
@@ -61,13 +71,24 @@ public void isWsdlTest()
 
     @Test
-    public void isXmlContentType() throws IOException {
+    public void ShouldNotAlertWithJustXmlContentType() throws IOException {
```
The leading S should be in lowercase, same for the other test.

ricekot replied:
You could also add a third test using `Sample#setSoapVersionTwoRequest()` for `application/soap+xml`.

ricekot replied:
You could also add a third test for the `application/wsdl+xml` content type...

thc202 commented on addOns/soap/CHANGELOG.md:
```diff
@@ -26,6 +26,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 - Fix exception with Java 9+ (Issue 4037).
 - SOAP operations are no longer overwritten in sites tree (Issue 1867).
 - Persist the add-on configuration required by the scan rules in the ZAP database (Issue 4866).
+- Fix detection of WSDL files (Issue 6440).
```
This needs to be added to the unreleased section.

theamanrawat replied:
Done.

thc202 commented on addOns/soap/CHANGELOG.md:
```diff
@@ -4,7 +4,7 @@ All notable changes to this add-on will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 ## Unreleased
-
```
`### Fixed`

theamanrawat replied:
> The commits can be squashed.

How can I do that?

thc202 commented on addOns/soap/src/test/java/org/zaproxy/zap/extension/soap/WSDLFilePassiveScanRuleTestCase.java:
```diff
@@ -22,14 +22,28 @@
 public class WSDLFilePassiveScanRuleTestCase {
     private HttpMessage wsdlMsg = new HttpMessage();
 
+    private HttpMessage setContentType(HttpMessage msg, String contentType)
+            throws HttpMalformedHeaderException {
+        msg.setResponseHeader(
+                "HTTP/1.1 200 OK\r\n"
+                        + "Date: Sun, 10 Aug 2014 15:35:20 GMT\r\n"
+                        + "Server: Apache/2.4.3 (Win32) OpenSSL/1.0.1c PHP/5.4.7\r\n"
+                        + "Content-Type: "
+                        + contentType
+                        + "\r\n");
+        return msg;
```
```Java
private static void setContentType(HttpMessage msg, String contentType){
    msg.getResponseHeader().setHeader(HttpHeader.CONTENT_TYPE, contentType);
}
```

thc202 commented:
The commits can be squashed.

theamanrawat commented:
> The commits can be squashed.

How to squash commits?

I tried to squash but it is giving me an error

```
error: cannot 'squash' without a previous commit
```

theamanrawat commented:
I think I did something wrong. How can I fix this please help.

ricekot commented:
Try doing
```bash
git pull --rebase upstream master
git push -f
```
on this branch.

theamanrawat commented:
> Try doing
> 
> ```shell
> git pull --rebase upstream master
> git push -f
> ```
> 
> on this branch.

Thank you so much

thc202 commented:
Thank you!

theamanrawat commented:
Your welcome.
